File: queues/ltcd_239.py
'''
LeetCode - problem 239
'''

from collections import deque
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        if k > len(nums) or not nums:
            return []
        
        ret = []
        q = deque()
        
        for i in range(len(nums)):
            while q and q[-1][0] <= nums[i]:
                q.pop()
            q.append((nums[i], i))
            while q[0][1] <= i - k:
                q.popleft()
            if i >= k - 1:
                ret.append(q[0][0])
        
        return ret

File: queues/ltcd_239.cpp
//LeetCode - problem 239

   vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int sum = 0;
        priority_queue<pair<int, int>> heap;
	   for(int i = 0; i < k-1;i++){
            heap.push({nums[i], i + k - 1});
        }
        vector<int> ans;
        for(int i = k-1;i < nums.size();i++){
            heap.push({nums[i], i + k - 1});
            while(heap.top().second < i){
                heap.pop();
            }
            ans.push_back(heap.top().first);
        }
        return ans;
    }
